From a15a12ec238dcf0066a648450d661c9a07986852 Mon Sep 17 00:00:00 2001
From: Steven Lee <steven_lee@aspeedtech.com>
Date: Mon, 6 Feb 2023 15:55:13 +0800
Subject: [PATCH] Potentially connect udc device incorrectly

The fs::directory_iterator doesn't specify the order of enumeration of
folders.  The if-condition here doesn't prevent the folder other than
gadget port, e.g. 1e6a0000.usb-vhub:pX, be taken into account, e.g.
power folder.  In this case, an exception will be generate internally
in connect().  This patch add another check to make sure the folder is
a gadget port.

Tested with evb-ast2500 to connect VM, then launch KVM and disconnect
for several times.

Fixes: c11257d864ff ("Connect HID gadget device dynamically")
Signed-off-by: Troy Lee <troy_lee@aspeedtech.com>
Change-Id: I0cad041d69d580a544e60c79e1c991482ca79819
---
 ikvm_input.cpp | 1 +
 1 file changed, 1 insertion(+)

diff --git a/ikvm_input.cpp b/ikvm_input.cpp
index 107756b..7a63142 100644
--- a/ikvm_input.cpp
+++ b/ikvm_input.cpp
@@ -58,6 +58,7 @@ void Input::connect()
             for (const auto& port : fs::directory_iterator(usbVirtualHubPath))
             {
                 if (fs::is_directory(port) && !fs::is_symlink(port) &&
+                    fs::exists(port.path() / "gadget") &&
                     !fs::exists(port.path() / "gadget/suspended"))
                 {
                     const std::string portId = port.path().filename();
-- 
2.25.1

